text/0000-proc-macro-expansion-api.md

```diff
@@ -0,0 +1,411 @@
+
+### Calling Scopes
+
+The method `call_from` sets the calling scope for the macro. What does this mean?
+
+Say we are defining a macro `my_proc!` and want to use another macro `helper!` as part of `my_proc!`. If `helper!` is hygienic, then all of its new variables and modules and whatever will live in its own `[Def]` scope independent the `[Def]` scope of `my_proc!`.
+
+If `helper!` is _unhygienic_ then any unhygienic declarations will live in the `[Call]` scope of `helper!` - but which scope is that? Assume that `helper!` expands to something like this:
+
+```rust
+struct S; // [Def]
+struct T; // [Call]
+
+// [Call]
+//   v
+impl T {
+    // These implementation functions can refer to S because
+    // they're in the same scope
+    ... // [Def]
+}
+```
+
+* If the `[Call]` scope of `helper!` is the `[Def]` scope of `my_proc!`, then `helper!` will 'export' or 'expose' the declaration of `T` to `my_proc!`, which lets `my_proc!` refer to `T`. This lets us delegate part of the implementation of `my_proc!` to other proc and decl macros (perhaps from other crates).
+
+* If instead the `[Call]` scope of `helper!` is the `[Call]` scope of `my_proc!`, then `helper!` will export the declarations to the caller of `my_proc!` instead of `my_proc!`. If we don't need access to `T` and just want to export it straight to the caller of `my_proc!` (or if `helper!` is actually just part of the caller's input to `my_proc!`, like `my_proc!(helper!(...))`) then this is what we want.
+
+Since both of these are legitimate use cases, `MacroCall` provides `call_from` to set what the `[Call]` scope of the macro call will be.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The proposed additions to the proc macro API in `proc_macro` are outlined above in the [API overview](#api-overview). Here we focus on technical challenges.
+
+When a source file is parsed any `macro_rules!` and `macro` definitions get added to a definition map long before the first macro is expanded. Procedural macros currently need to live in a separate crate, and it seems they will for a while. This means that _in principle_ any macro call that would resolve in the caller's scope should be available to resolve at the time the proc macro is expanded.
+
+Built-in macros already look more and more like proc macros (or at the very least could be massaged into acting like them), and so they can also be added to the definition map.
+
+Since proc macros and `macro` definitions are relative-path-addressable, the proc macro call context needs to keep track of what the path was at the call site. I'm not sure if this information is available at expansion time, but are there any issues getting it?
```

Comment:
Yeah, this information is available at expansion time. Resolving the macro shouldn't be a problem; I can write up notes on how to implement.